src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -679,6 +691,37 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
 
             if (feature?.CanHaveBody == true)
             {
+                if (isRawBodyType)
+                {
+                    var bodyReader = httpContext.Request.BodyReader;
+
+                    while (true)
+                    {
+                        var result = await bodyReader.ReadAsync();
+                        var buffer = result.Buffer;
+
+                        if (result.IsCompleted)
+                        {
+                            if (bodyType == typeof(ReadOnlySequence<byte>))
+                            {
+                                // REVIEW: Does this need to be a copy? We can tell users to consume the buffer
+                                // immediately in the action
+                                return (buffer, true);
```

Comment:
When is `PipeReader.AdvanceTo` called to say the body is consumed? Before or after the request delegate is executed?

Reply:
After thinking about this, it's actually broken for non Kestrel servers since nobody consumes that pipe currently. I'll need to add more logic there.

Reply:
Hmm actually no, it seems like we complete the pipe reader in both cases https://github.com/dotnet/aspnetcore/blob/694507a5d24fd858dff8231829e42fb21acbe9b9/src/Http/Http/src/Features/RequestBodyPipeFeature.cs#L35-L48